### ISSUE TITLE
QGIS Warning and python's errors with Linking relation editor 

### ISSUE THREAD

**Issue by pablolivares** [opengisch/qgis-linking-relation-editor#43]
We have a Transport public model with different tables and relations. When we tried to edit some of the features we have some QGIS Warnings  which are not very reassuring. You can check the model example here [Transports_publics_TEST.zip](https://github.com/opengisch/qgis-linking-relation-editor/files/12522063/Transports_publics_TEST.zip)
  I'll try to explain in best way some of our issues. 

First, we have some intermediate tables for the relations N to M. One of the problems we found was when opening one of the intermediate tables (link_troncon_ligneoffre). This warning appears, saying that some relationships are missing in the configuration

![warning1](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/fbd8da89-cf0e-4bbc-b1e4-1917ab4c4216)

However, these relationships are established in the project properties as shown in the following image

![relations](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/e8a6d05f-d4aa-45f5-b690-929d6b736136)

A second problem we've found is when we try to associate several sections with a new offer line. We followed the next steps

-Creation of a new service line called "Lajoux-Lajoux"
![nouvelle ligne](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/2cfa9c9a-22c2-4082-92db-db7093de0d81)

-We have selected the sections to be edited and through the offer line tab we have made the link with the new "Lajoux-Lajoux" offer line.
![lien ligne_troncon](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/774e627d-c088-4d6e-9862-1632daf1c3b7)
![lien ligne_troncon_2](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/142b95a3-3ff7-4659-8071-99873f9d0104)

But once you've clicked on Ok, a Warning appears, saying that a Python error has occurred

![erreur python](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/aa3cebe2-4a86-459b-b1de-b8a444590954)

On the other hand, if we consult the intermediate table between Sections and Offer Lines (link_troncon_offerline) we notice that it is updated with the new relationships (new offer line ID= 40, selected sections ID= 374, 392, 393, 394, 508).

![table_interm](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/12312c83-0dd0-40ed-afbe-c433aa8a6093)

Now, we are migrating to PostGIS, and we would know how to manage these issues. Also we would like to know if in PostGIS we have to establish the relations between the tables or we have to do it just in QGIS.

thanks in advance and don't hesitate to contact us if you have questions about it

**Comment by domi4484** [comment]
@pablolivares Thank you for the report!

The first issue doesn't seem related to the plugin, but probably to the QGIS Relation Reference widget.
The relations seems configured correctly to me. I noticed, that if i change the widget configuration for `Troncon_id` and `LigneOffre_id` from `Relation Reference` to `Text Edit`, and back to `Relation Reference` the warnings are not generated anymore. Maybe during time you changed something in the relations and the widget keeps a reference to an old relation. You could open a bug in QGIS for that.
 
![image](https://github.com/opengisch/qgis-linking-relation-editor/assets/9881900/335493c3-a5d9-4cef-90c4-d83bb97e20bd)


The second problem is because the Linking Relation editor doesn't support yet multiedit (linking one item for multiple selected parent features). I'll see if I can replace the ugly python exception with a nicer dialog telling that.
If you are interested in the multiedit feature for this widget you can ask for an offer at [info@opengis.ch](mailto:info@opengis.ch) or provide a pull request to this repository.

**Comment by pablolivares** [comment]
@domi4484 thank you for answer. The first issue it's also ok when I changed the widget configuration, thank you again. 

So, the multiedit it's not supported and we'll have this warning/error. But, what about the last image what I posted? It's seems like it's already working, even if we have this warning/error, because one item (ligne d'offre) was linked succesfully to multiple selected parent features (tronçons), at leats in the intermediate table.
![table_interm](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/42aae4a9-3b9f-4193-ae9b-261c0f70ba3f)

**Comment by domi4484** [comment]
@pablolivares oh I have overseen the last part of your report. 

Looking in the code it seems I had implemented some parts of the multiedit, but it was never tested. So it seems that the "multi" links are done correctly, but the dialog has a problem when reloading the UI to display them.
More precisely at this place apparently a `QTreeWidget` can't be used as key for a `dict` and the implementation should be changed a little:
```python
if treeWidgetItem in multimapChildFeatures:
```


About the second question, I would create the relations in PostGIS, so the data can be handled correctly also outside the context of QGIS.
Btw. in Project -> Properties -> Relations there is a button to discover automatically the relations from the database, so you don't have to define them twice (I think it works only with postgis).
![image](https://github.com/opengisch/qgis-linking-relation-editor/assets/9881900/878008c7-98c0-4e38-a934-436b21a744d0)

**Comment by pablolivares** [comment]
@domi4484  Thanks for the feedback. I made the relations in postgis with PG ADMIN 4. For the intermediates tables, I did it with this code to stablish a Composition relations:

```
CONSTRAINT link_arret_ligne_offre_objectid_arret_fkey FOREIGN KEY (objectid_arret)
        REFERENCES sdt_transports.arret (objectid_arret) MATCH SIMPLE
        ON UPDATE CASCADE
        ON DELETE CASCADE
        NOT VALID,
    CONSTRAINT link_arret_ligne_offre_objectid_ligne_offre_fkey FOREIGN KEY (objectid_ligne_offre)
        REFERENCES sdt_transports.ligne_offre (objectid_ligne_offre) MATCH SIMPLE
        ON UPDATE CASCADE
        ON DELETE CASCADE
        NOT VALID
```

But in QGIS with the discover relations we get only Association relations. 
![image](https://github.com/opengisch/qgis-linking-relation-editor/assets/144106571/429d45c9-512a-4563-9276-082af9b27656)

Do you know if there is some method to stablish composition relations in PG, or if its enough with the code used for the editions or multieditions in QGIS?

Thanks in advance

Pablo

**Comment by domi4484** [comment]
For what I could quickly look int the QGIS source code, there is no code to decide if the discovered relation should be an association or composition. And by default it is association.

So I fear in that case you will have to create those relations by hand, because unfortunately it seems not possible to change the "Force" after the relation has been created.

**Comment by domi4484** [comment]
@pablolivares lately I made some developement related to multiedit published in version 1.3.0.
I'll close this issue, please feel free to open a new one if there are still problems
